File: packages/nerva-py/nerva/runtime/circuit_breaker.py
"""Circuit breaker — prevent cascading failures by tracking handler health.

State machine:
    CLOSED  (normal)  -- allow calls, count consecutive failures
        |  failure_threshold exceeded
        v
    OPEN    (failing) -- reject all calls
        |  recovery_seconds elapsed
        v
    HALF_OPEN (testing) -- allow limited test calls
        |  success -> CLOSED  |  failure -> OPEN
"""

from __future__ import annotations

import threading
import time
from dataclasses import dataclass
from enum import StrEnum

__all__ = ["CircuitState", "CircuitBreakerConfig", "CircuitBreaker"]


class CircuitState(StrEnum):
    """States in the circuit breaker state machine.

    Members:
        CLOSED: Normal operation, calls flow through.
        OPEN: Handler is failing, calls are rejected.
        HALF_OPEN: Recovery probe, limited calls allowed.
    """

    CLOSED = "closed"
    OPEN = "open"
    HALF_OPEN = "half_open"


# ---------------------------------------------------------------------------
# Named constants
# ---------------------------------------------------------------------------

DEFAULT_FAILURE_THRESHOLD = 3
"""Consecutive failures before the circuit opens."""

DEFAULT_RECOVERY_SECONDS = 60.0
"""Seconds after opening before transitioning to half-open."""

DEFAULT_HALF_OPEN_MAX_CALLS = 1
"""Number of probe calls allowed in the half-open state."""


@dataclass(frozen=True)
class CircuitBreakerConfig:
    """Tunable thresholds for a circuit breaker.

    Attributes:
        failure_threshold: Consecutive failures before the circuit opens.
        recovery_seconds: Seconds to wait before allowing a recovery probe.
        half_open_max_calls: Maximum probe calls permitted in half-open state.
    """

    failure_threshold: int = DEFAULT_FAILURE_THRESHOLD
    recovery_seconds: float = DEFAULT_RECOVERY_SECONDS
    half_open_max_calls: int = DEFAULT_HALF_OPEN_MAX_CALLS


class CircuitBreaker:
    """Per-handler circuit breaker with a thread-safe state machine.

    Tracks success/failure for a specific handler. Opens the circuit after
    ``failure_threshold`` consecutive failures, then transitions to half-open
    after ``recovery_seconds`` to test whether the handler has recovered.

    Example::

        breaker = CircuitBreaker()
        if not breaker.is_allowed():
            raise RuntimeError("circuit open")
        try:
            result = await call_handler()
            breaker.record_success()
        except Exception:
            breaker.record_failure()
            raise
    """

    def __init__(self, config: CircuitBreakerConfig | None = None) -> None:
        """Create a new circuit breaker in the CLOSED state.

        Args:
            config: Optional thresholds. Uses defaults when None.
        """
        self._config = config or CircuitBreakerConfig()
        self._lock = threading.Lock()
        self._state = CircuitState.CLOSED
        self._consecutive_failures = 0
        self._opened_at = 0.0
        self._half_open_calls = 0

    # -- public read-only properties ----------------------------------------

    @property
    def state(self) -> CircuitState:
        """Return the current circuit state, applying time-based transitions.

        Returns:
            The live circuit state after evaluating recovery timeout.
        """
        with self._lock:
            self._maybe_transition_to_half_open()
            return self._state

    # -- public mutation methods --------------------------------------------

    def is_allowed(self) -> bool:
        """Check if a call is allowed under the current circuit state.

        Returns:
            True if the call may proceed, False if the circuit is open.
        """
        with self._lock:
            self._maybe_transition_to_half_open()
            return self._is_allowed_locked()

    def record_success(self) -> None:
        """Record a successful call.

        Resets the failure counter. If the circuit is half-open, closes it.
        """
        with self._lock:
            self._consecutive_failures = 0
            self._half_open_calls = 0

            if self._state == CircuitState.HALF_OPEN:
                self._state = CircuitState.CLOSED

    def record_failure(self) -> None:
        """Record a failed call.

        Increments the consecutive failure counter. Opens the circuit when
        the failure threshold is reached. Re-opens immediately if called
        in the half-open state.
        """
        with self._lock:
            self._consecutive_failures += 1

            if self._state == CircuitState.HALF_OPEN:
                self._open_locked()
                return

            if self._consecutive_failures >= self._config.failure_threshold:
                self._open_locked()

    def reset(self) -> None:
        """Force-reset the circuit to the CLOSED state.

        Clears all failure counts and timers.
        """
        with self._lock:
            self._state = CircuitState.CLOSED
            self._consecutive_failures = 0
            self._opened_at = 0.0
            self._half_open_calls = 0

    # -- private helpers (must be called with _lock held) -------------------

    def _open_locked(self) -> None:
        """Transition to OPEN and record the opening timestamp."""
        self._state = CircuitState.OPEN
        self._opened_at = time.monotonic()
        self._half_open_calls = 0

    def _maybe_transition_to_half_open(self) -> None:
        """Move from OPEN to HALF_OPEN if recovery_seconds have elapsed."""
        if self._state != CircuitState.OPEN:
            return
        elapsed = time.monotonic() - self._opened_at
        if elapsed >= self._config.recovery_seconds:
            self._state = CircuitState.HALF_OPEN
            self._half_open_calls = 0

    def _is_allowed_locked(self) -> bool:
        """Evaluate whether a call is permitted under current state."""
        if self._state == CircuitState.CLOSED:
            return True

        if self._state == CircuitState.OPEN:
            return False

        # HALF_OPEN: allow up to half_open_max_calls probe calls
        if self._half_open_calls < self._config.half_open_max_calls:
            self._half_open_calls += 1
            return True

        return False

File: packages/nerva-js/src/runtime/circuit-breaker.ts
/**
 * Circuit breaker — prevent cascading failures by tracking handler health.
 *
 * State machine:
 *   CLOSED  (normal)  -- allow calls, count consecutive failures
 *       |  failureThreshold exceeded
 *       v
 *   OPEN    (failing) -- reject all calls
 *       |  recoveryMs elapsed
 *       v
 *   HALF_OPEN (testing) -- allow limited test calls
 *       |  success -> CLOSED  |  failure -> OPEN
 *
 * @module runtime/circuit-breaker
 */

// ---------------------------------------------------------------------------
// CircuitState
// ---------------------------------------------------------------------------

/**
 * States in the circuit breaker state machine.
 */
export const CircuitState = {
  /** Normal operation, calls flow through. */
  CLOSED: "closed",
  /** Handler is failing, calls are rejected. */
  OPEN: "open",
  /** Recovery probe, limited calls allowed. */
  HALF_OPEN: "half_open",
} as const;

export type CircuitState = (typeof CircuitState)[keyof typeof CircuitState];

// ---------------------------------------------------------------------------
// Named constants
// ---------------------------------------------------------------------------

/** Consecutive failures before the circuit opens. */
const DEFAULT_FAILURE_THRESHOLD = 3;

/** Milliseconds after opening before transitioning to half-open. */
const DEFAULT_RECOVERY_MS = 60_000;

/** Number of probe calls allowed in the half-open state. */
const DEFAULT_HALF_OPEN_MAX_CALLS = 1;

// ---------------------------------------------------------------------------
// CircuitBreakerConfig
// ---------------------------------------------------------------------------

/**
 * Tunable thresholds for a circuit breaker.
 */
export interface CircuitBreakerConfig {
  /** Consecutive failures before the circuit opens. */
  readonly failureThreshold: number;
  /** Milliseconds to wait before allowing a recovery probe. */
  readonly recoveryMs: number;
  /** Maximum probe calls permitted in half-open state. */
  readonly halfOpenMaxCalls: number;
}

/**
 * Create a CircuitBreakerConfig with defaults for any omitted fields.
 *
 * @param overrides - Partial config to merge with defaults.
 * @returns A complete CircuitBreakerConfig.
 */
export function createCircuitBreakerConfig(
  overrides?: Partial<CircuitBreakerConfig>,
): CircuitBreakerConfig {
  return {
    failureThreshold: overrides?.failureThreshold ?? DEFAULT_FAILURE_THRESHOLD,
    recoveryMs: overrides?.recoveryMs ?? DEFAULT_RECOVERY_MS,
    halfOpenMaxCalls: overrides?.halfOpenMaxCalls ?? DEFAULT_HALF_OPEN_MAX_CALLS,
  };
}

// ---------------------------------------------------------------------------
// Now provider (injectable for testing)
// ---------------------------------------------------------------------------

/**
 * Returns the current monotonic-ish timestamp in milliseconds.
 * Defaults to `performance.now()` but can be overridden for testing.
 */
type NowProvider = () => number;

// ---------------------------------------------------------------------------
// CircuitBreaker
// ---------------------------------------------------------------------------

/**
 * Per-handler circuit breaker with a state machine.
 *
 * Tracks success/failure for a specific handler. Opens the circuit after
 * `failureThreshold` consecutive failures, then transitions to half-open
 * after `recoveryMs` to test whether the handler has recovered.
 *
 * @example
 * ```ts
 * const breaker = new CircuitBreaker();
 * if (!breaker.isAllowed()) {
 *   throw new Error("circuit open");
 * }
 * try {
 *   const result = await callHandler();
 *   breaker.recordSuccess();
 * } catch {
 *   breaker.recordFailure();
 *   throw;
 * }
 * ```
 */
export class CircuitBreaker {
  private readonly _config: CircuitBreakerConfig;
  private readonly _now: NowProvider;
  private _state: CircuitState;
  private _consecutiveFailures: number;
  private _openedAt: number;
  private _halfOpenCalls: number;

  /**
   * Create a new circuit breaker in the CLOSED state.
   *
   * @param config - Optional thresholds. Uses defaults when omitted.
   * @param nowProvider - Optional time source for testing. Defaults to `performance.now`.
   */
  constructor(config?: Partial<CircuitBreakerConfig>, nowProvider?: NowProvider) {
    this._config = createCircuitBreakerConfig(config);
    this._now = nowProvider ?? (() => performance.now());
    this._state = CircuitState.CLOSED;
    this._consecutiveFailures = 0;
    this._openedAt = 0;
    this._halfOpenCalls = 0;
  }

  /**
   * Return the current circuit state, applying time-based transitions.
   *
   * @returns The live circuit state after evaluating recovery timeout.
   */
  get state(): CircuitState {
    this._maybeTransitionToHalfOpen();
    return this._state;
  }

  /**
   * Check if a call is allowed under the current circuit state.
   *
   * @returns `true` if the call may proceed, `false` if the circuit is open.
   */
  isAllowed(): boolean {
    this._maybeTransitionToHalfOpen();
    return this._isAllowedInternal();
  }

  /**
   * Record a successful call.
   *
   * Resets the failure counter. If the circuit is half-open, closes it.
   */
  recordSuccess(): void {
    this._consecutiveFailures = 0;
    this._halfOpenCalls = 0;

    if (this._state === CircuitState.HALF_OPEN) {
      this._state = CircuitState.CLOSED;
    }
  }

  /**
   * Record a failed call.
   *
   * Increments the consecutive failure counter. Opens the circuit when
   * the failure threshold is reached. Re-opens immediately if called
   * in the half-open state.
   */
  recordFailure(): void {
    this._consecutiveFailures += 1;

    if (this._state === CircuitState.HALF_OPEN) {
      this._open();
      return;
    }

    if (this._consecutiveFailures >= this._config.failureThreshold) {
      this._open();
    }
  }

  /**
   * Force-reset the circuit to the CLOSED state.
   *
   * Clears all failure counts and timers.
   */
  reset(): void {
    this._state = CircuitState.CLOSED;
    this._consecutiveFailures = 0;
    this._openedAt = 0;
    this._halfOpenCalls = 0;
  }

  // -- Private helpers ------------------------------------------------------

  /** Transition to OPEN and record the opening timestamp. */
  private _open(): void {
    this._state = CircuitState.OPEN;
    this._openedAt = this._now();
    this._halfOpenCalls = 0;
  }

  /** Move from OPEN to HALF_OPEN if recoveryMs have elapsed. */
  private _maybeTransitionToHalfOpen(): void {
    if (this._state !== CircuitState.OPEN) return;

    const elapsed = this._now() - this._openedAt;
    if (elapsed >= this._config.recoveryMs) {
      this._state = CircuitState.HALF_OPEN;
      this._halfOpenCalls = 0;
    }
  }

  /** Evaluate whether a call is permitted under current state. */
  private _isAllowedInternal(): boolean {
    if (this._state === CircuitState.CLOSED) return true;
    if (this._state === CircuitState.OPEN) return false;

    // HALF_OPEN: allow up to halfOpenMaxCalls probe calls
    if (this._halfOpenCalls < this._config.halfOpenMaxCalls) {
      this._halfOpenCalls += 1;
      return true;
    }

    return false;
  }
}
